// @docusaurus/theme-classic/lib/theme/DocVersionBanner/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _useDocusaurusContext = _interopRequireDefault(require("@docusaurus/useDocusaurusContext"));

var _Link = _interopRequireDefault(require("@docusaurus/Link"));

var _Translate = _interopRequireDefault(require("@docusaurus/Translate"));

var _useDocs = require("@theme/hooks/useDocs");

var _themeCommon = require("@docusaurus/theme-common");

var _clsx = _interopRequireDefault(require("clsx"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function UnreleasedVersionLabel({
  siteTitle,
  versionMetadata
}) {
  return <_Translate.default id="theme.docs.versions.unreleasedVersionLabel" description="The label used to tell the user that he's browsing an unreleased doc version" values={{
    siteTitle,
    versionLabel: <b>{versionMetadata.label}</b>
  }}>
      {'This is unreleased documentation for {siteTitle} {versionLabel} version.'}
    </_Translate.default>;
}

function UnmaintainedVersionLabel({
  siteTitle,
  versionMetadata
}) {
  return <_Translate.default id="theme.docs.versions.unmaintainedVersionLabel" description="The label used to tell the user that he's browsing an unmaintained doc version" values={{
    siteTitle,
    versionLabel: <b>{versionMetadata.label}</b>
  }}>
      {'This is documentation for {siteTitle} {versionLabel}, which is no longer actively maintained.'}
    </_Translate.default>;
}

const BannerLabelComponents = {
  unreleased: UnreleasedVersionLabel,
  unmaintained: UnmaintainedVersionLabel
};

function BannerLabel(props) {
  const BannerLabelComponent = BannerLabelComponents[props.versionMetadata.banner];
  return <BannerLabelComponent {...props} />;
}

function LatestVersionSuggestionLabel({
  versionLabel,
  to,
  onClick
}) {
  return <_Translate.default id="theme.docs.versions.latestVersionSuggestionLabel" description="The label used to tell the user to check the latest version" values={{
    versionLabel,
    latestVersionLink: <b>
            <_Link.default to={to} onClick={onClick}>
              <_Translate.default id="theme.docs.versions.latestVersionLinkLabel" description="The label used for the latest version suggestion link label">
                latest version
              </_Translate.default>
            </_Link.default>
          </b>
  }}>
      {'For up-to-date documentation, see the {latestVersionLink} ({versionLabel}).'}
    </_Translate.default>;
}

function DocVersionBannerEnabled({
  versionMetadata
}) {
  const {
    siteConfig: {
      title: siteTitle
    }
  } = (0, _useDocusaurusContext.default)();
  const {
    pluginId
  } = (0, _useDocs.useActivePlugin)({
    failfast: true
  });

  const getVersionMainDoc = version => version.docs.find(doc => doc.id === version.mainDocId);

  const {
    savePreferredVersionName
  } = (0, _themeCommon.useDocsPreferredVersion)(pluginId);
  const {
    latestDocSuggestion,
    latestVersionSuggestion
  } = (0, _useDocs.useDocVersionSuggestions)(pluginId); // try to link to same doc in latest version (not always possible)
  // fallback to main doc of latest version

  const latestVersionSuggestedDoc = latestDocSuggestion !== null && latestDocSuggestion !== void 0 ? latestDocSuggestion : getVersionMainDoc(latestVersionSuggestion);
  return <div className={(0, _clsx.default)(_themeCommon.ThemeClassNames.docs.docVersionBanner, 'alert alert--warning margin-bottom--md')} role="alert">
      <div>
        <BannerLabel siteTitle={siteTitle} versionMetadata={versionMetadata} />
      </div>
      <div className="margin-top--md">
        <LatestVersionSuggestionLabel versionLabel={latestVersionSuggestion.label} to={latestVersionSuggestedDoc.path} onClick={() => savePreferredVersionName(latestVersionSuggestion.name)} />
      </div>
    </div>;
}

function DocVersionBanner({
  versionMetadata
}) {
  if (versionMetadata.banner) {
    return <DocVersionBannerEnabled versionMetadata={versionMetadata} />;
  }

  return <></>;
}

var _default = DocVersionBanner;
exports.default = _default;
