// @docusaurus/theme-classic/lib/theme/Tabs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useUserPreferencesContext = _interopRequireDefault(require("@theme/hooks/useUserPreferencesContext"));

var _clsx = _interopRequireDefault(require("clsx"));

var _stylesModule = _interopRequireDefault(require("./styles.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function isInViewport(element) {
  const {
    top,
    left,
    bottom,
    right
  } = element.getBoundingClientRect();
  const {
    innerHeight,
    innerWidth
  } = window;
  return top >= 0 && right <= innerWidth && bottom <= innerHeight && left >= 0;
}

function Tabs(props) {
  var _children$find;

  const {
    lazy,
    block,
    defaultValue: defaultValueProp,
    values: valuesProp,
    groupId,
    className
  } = props;

  const children = _react.Children.toArray(props.children);

  const values = valuesProp !== null && valuesProp !== void 0 ? valuesProp : children.map(child => {
    return {
      value: child.props.value,
      label: child.props.label
    };
  });
  const defaultValue = defaultValueProp !== null && defaultValueProp !== void 0 ? defaultValueProp : (_children$find = children.find(child => child.props.default)) === null || _children$find === void 0 ? void 0 : _children$find.props.value;
  const {
    tabGroupChoices,
    setTabGroupChoices
  } = (0, _useUserPreferencesContext.default)();
  const [selectedValue, setSelectedValue] = (0, _react.useState)(defaultValue);
  const tabRefs = [];

  if (groupId != null) {
    const relevantTabGroupChoice = tabGroupChoices[groupId];

    if (relevantTabGroupChoice != null && relevantTabGroupChoice !== selectedValue && values.some(value => value.value === relevantTabGroupChoice)) {
      setSelectedValue(relevantTabGroupChoice);
    }
  }

  const handleTabChange = event => {
    const selectedTab = event.currentTarget;
    const selectedTabIndex = tabRefs.indexOf(selectedTab);
    const selectedTabValue = values[selectedTabIndex].value;
    setSelectedValue(selectedTabValue);

    if (groupId != null) {
      setTabGroupChoices(groupId, selectedTabValue);
      setTimeout(() => {
        if (isInViewport(selectedTab)) {
          return;
        }

        selectedTab.scrollIntoView({
          block: 'center',
          behavior: 'smooth'
        });
        selectedTab.classList.add(_stylesModule.default.tabItemActive);
        setTimeout(() => selectedTab.classList.remove(_stylesModule.default.tabItemActive), 2000);
      }, 150);
    }
  };

  const handleKeydown = event => {
    var _focusElement;

    let focusElement = null;

    switch (event.key) {
      case 'ArrowRight':
        {
          const nextTab = tabRefs.indexOf(event.target) + 1;
          focusElement = tabRefs[nextTab] || tabRefs[0];
          break;
        }

      case 'ArrowLeft':
        {
          const prevTab = tabRefs.indexOf(event.target) - 1;
          focusElement = tabRefs[prevTab] || tabRefs[tabRefs.length - 1];
          break;
        }

      default:
        break;
    }

    (_focusElement = focusElement) === null || _focusElement === void 0 ? void 0 : _focusElement.focus();
  };

  return <div className="tabs-container">
      <ul role="tablist" aria-orientation="horizontal" className={(0, _clsx.default)('tabs', {
      'tabs--block': block
    }, className)}>
        {values.map(({
        value,
        label
      }) => <li role="tab" tabIndex={selectedValue === value ? 0 : -1} aria-selected={selectedValue === value} className={(0, _clsx.default)('tabs__item', _stylesModule.default.tabItem, {
        'tabs__item--active': selectedValue === value
      })} key={value} ref={tabControl => tabRefs.push(tabControl)} onKeyDown={handleKeydown} onFocus={handleTabChange} onClick={handleTabChange}>
            {label !== null && label !== void 0 ? label : value}
          </li>)}
      </ul>

      {lazy ? (0, _react.cloneElement)(children.filter(tabItem => tabItem.props.value === selectedValue)[0], {
      className: 'margin-vert--md'
    }) : <div className="margin-vert--md">
          {children.map((tabItem, i) => (0, _react.cloneElement)(tabItem, {
        key: i,
        hidden: tabItem.props.value !== selectedValue
      }))}
        </div>}
    </div>;
}

var _default = Tabs;
exports.default = _default;
