// @docusaurus/plugin-content-docs/lib/docs.js
"use strict";
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.processDocMetadata = exports.readVersionDocs = exports.readDocFile = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const utils_1 = require("@docusaurus/utils");
const lastUpdate_1 = require("./lastUpdate");
const slug_1 = tslib_1.__importDefault(require("./slug"));
const constants_1 = require("./constants");
const versions_1 = require("./versions");
const numberPrefix_1 = require("./numberPrefix");
const docFrontMatter_1 = require("./docFrontMatter");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
async function readLastUpdateData(filePath, options) {
    const { showLastUpdateAuthor, showLastUpdateTime } = options;
    if (showLastUpdateAuthor || showLastUpdateTime) {
        // Use fake data in dev for faster development.
        const fileLastUpdateData = process.env.NODE_ENV === 'production'
            ? await lastUpdate_1.getFileLastUpdate(filePath)
            : {
                author: 'Author',
                timestamp: 1539502055,
            };
        if (fileLastUpdateData) {
            const { author, timestamp } = fileLastUpdateData;
            return {
                lastUpdatedAt: showLastUpdateTime ? timestamp : undefined,
                lastUpdatedBy: showLastUpdateAuthor ? author : undefined,
            };
        }
    }
    return {};
}
async function readDocFile(versionMetadata, source, options) {
    const contentPath = await utils_1.getFolderContainingFile(versions_1.getDocsDirPaths(versionMetadata), source);
    const filePath = path_1.default.join(contentPath, source);
    const [content, lastUpdate] = await Promise.all([
        fs_extra_1.default.readFile(filePath, 'utf-8'),
        readLastUpdateData(filePath, options),
    ]);
    return { source, content, lastUpdate, contentPath, filePath };
}
exports.readDocFile = readDocFile;
async function readVersionDocs(versionMetadata, options) {
    const sources = await utils_1.Globby(options.include, {
        cwd: versionMetadata.contentPath,
        ignore: options.exclude,
    });
    return Promise.all(sources.map((source) => readDocFile(versionMetadata, source, options)));
}
exports.readVersionDocs = readVersionDocs;
function doProcessDocMetadata({ docFile, versionMetadata, context, options, }) {
    var _a, _b, _c, _d, _e, _f;
    const { source, content, lastUpdate, contentPath, filePath } = docFile;
    const { homePageId } = options;
    const { siteDir, i18n } = context;
    const { frontMatter: unsafeFrontMatter, contentTitle, excerpt, } = utils_1.parseMarkdownString(content);
    const frontMatter = docFrontMatter_1.validateDocFrontMatter(unsafeFrontMatter);
    const { custom_edit_url: customEditURL, 
    // Strip number prefixes by default (01-MyFolder/01-MyDoc.md => MyFolder/MyDoc) by default,
    // but allow to disable this behavior with frontmatter
    parse_number_prefixes: parseNumberPrefixes = true, } = frontMatter;
    // ex: api/plugins/myDoc -> myDoc
    // ex: myDoc -> myDoc
    const sourceFileNameWithoutExtension = path_1.default.basename(source, path_1.default.extname(source));
    // ex: api/plugins/myDoc -> api/plugins
    // ex: myDoc -> .
    const sourceDirName = path_1.default.dirname(source);
    const { filename: unprefixedFileName, numberPrefix } = parseNumberPrefixes
        ? options.numberPrefixParser(sourceFileNameWithoutExtension)
        : { filename: sourceFileNameWithoutExtension, numberPrefix: undefined };
    const baseID = (_a = frontMatter.id) !== null && _a !== void 0 ? _a : unprefixedFileName;
    if (baseID.includes('/')) {
        throw new Error(`Document id "${baseID}" cannot include slash.`);
    }
    // For autogenerated sidebars, sidebar position can come from filename number prefix or frontmatter
    const sidebarPosition = (_b = frontMatter.sidebar_position) !== null && _b !== void 0 ? _b : numberPrefix;
    // TODO legacy retrocompatibility
    // The same doc in 2 distinct version could keep the same id,
    // we just need to namespace the data by version
    const versionIdPrefix = versionMetadata.versionName === constants_1.CURRENT_VERSION_NAME
        ? undefined
        : `version-${versionMetadata.versionName}`;
    // TODO legacy retrocompatibility
    // I think it's bad to affect the frontmatter id with the dirname?
    function computeDirNameIdPrefix() {
        if (sourceDirName === '.') {
            return undefined;
        }
        // Eventually remove the number prefixes from intermediate directories
        return parseNumberPrefixes
            ? numberPrefix_1.stripPathNumberPrefixes(sourceDirName, options.numberPrefixParser)
            : sourceDirName;
    }
    const unversionedId = [computeDirNameIdPrefix(), baseID]
        .filter(Boolean)
        .join('/');
    // TODO is versioning the id very useful in practice?
    // legacy versioned id, requires a breaking change to modify this
    const id = [versionIdPrefix, unversionedId].filter(Boolean).join('/');
    // TODO remove soon, deprecated homePageId
    const isDocsHomePage = unversionedId === (homePageId !== null && homePageId !== void 0 ? homePageId : '_index');
    if (frontMatter.slug && isDocsHomePage) {
        throw new Error(`The docs homepage (homePageId=${homePageId}) is not allowed to have a frontmatter slug=${frontMatter.slug} => you have to choose either homePageId or slug, not both`);
    }
    const docSlug = isDocsHomePage
        ? '/'
        : slug_1.default({
            baseID,
            dirName: sourceDirName,
            frontmatterSlug: frontMatter.slug,
            stripDirNumberPrefixes: parseNumberPrefixes,
            numberPrefixParser: options.numberPrefixParser,
        });
    // Note: the title is used by default for page title, sidebar label, pagination buttons...
    // frontMatter.title should be used in priority over contentTitle (because it can contain markdown/JSX syntax)
    const title = (_d = (_c = frontMatter.title) !== null && _c !== void 0 ? _c : contentTitle) !== null && _d !== void 0 ? _d : baseID;
    const description = (_f = (_e = frontMatter.description) !== null && _e !== void 0 ? _e : excerpt) !== null && _f !== void 0 ? _f : '';
    const permalink = utils_1.normalizeUrl([versionMetadata.versionPath, docSlug]);
    function getDocEditUrl() {
        const relativeFilePath = path_1.default.relative(contentPath, filePath);
        if (typeof options.editUrl === 'function') {
            return options.editUrl({
                version: versionMetadata.versionName,
                versionDocsDirPath: utils_1.posixPath(path_1.default.relative(siteDir, versionMetadata.contentPath)),
                docPath: utils_1.posixPath(relativeFilePath),
                permalink,
                locale: context.i18n.currentLocale,
            });
        }
        else if (typeof options.editUrl === 'string') {
            const isLocalized = contentPath === versionMetadata.contentPathLocalized;
            const baseVersionEditUrl = isLocalized && options.editLocalizedFiles
                ? versionMetadata.versionEditUrlLocalized
                : versionMetadata.versionEditUrl;
            return utils_1.getEditUrl(relativeFilePath, baseVersionEditUrl);
        }
        else {
            return undefined;
        }
    }
    // Assign all of object properties during instantiation (if possible) for
    // NodeJS optimization.
    // Adding properties to object after instantiation will cause hidden
    // class transitions.
    return {
        unversionedId,
        id,
        isDocsHomePage,
        title,
        description,
        source: utils_1.aliasedSitePath(filePath, siteDir),
        sourceDirName,
        slug: docSlug,
        permalink,
        editUrl: customEditURL !== undefined ? customEditURL : getDocEditUrl(),
        tags: utils_1.normalizeFrontMatterTags(versionMetadata.tagsPath, frontMatter.tags),
        version: versionMetadata.versionName,
        lastUpdatedBy: lastUpdate.lastUpdatedBy,
        lastUpdatedAt: lastUpdate.lastUpdatedAt,
        formattedLastUpdatedAt: lastUpdate.lastUpdatedAt
            ? new Intl.DateTimeFormat(i18n.currentLocale).format(lastUpdate.lastUpdatedAt * 1000)
            : undefined,
        sidebarPosition,
        frontMatter,
    };
}
function processDocMetadata(args) {
    try {
        return doProcessDocMetadata(args);
    }
    catch (e) {
        console.error(chalk_1.default.red(`Can't process doc metadatas for doc at path "${args.docFile.filePath}" in version "${args.versionMetadata.versionName}"`));
        throw e;
    }
}
exports.processDocMetadata = processDocMetadata;
