// @docusaurus/theme-classic/lib/theme/DocSidebar/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DocSidebar;

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _themeCommon = require("@docusaurus/theme-common");

var _useWindowSize = _interopRequireDefault(require("@theme/hooks/useWindowSize"));

var _useScrollPosition = _interopRequireDefault(require("@theme/hooks/useScrollPosition"));

var _Logo = _interopRequireDefault(require("@theme/Logo"));

var _IconArrow = _interopRequireDefault(require("@theme/IconArrow"));

var _Translate = require("@docusaurus/Translate");

var _DocSidebarItem = require("@theme/DocSidebarItem");

var _stylesModule = _interopRequireDefault(require("./styles.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function useShowAnnouncementBar() {
  const {
    isClosed
  } = (0, _themeCommon.useAnnouncementBar)();
  const [showAnnouncementBar, setShowAnnouncementBar] = (0, _react.useState)(!isClosed);
  (0, _useScrollPosition.default)(({
    scrollY
  }) => {
    if (!isClosed) {
      setShowAnnouncementBar(scrollY === 0);
    }
  });
  return showAnnouncementBar;
}

function HideableSidebarButton({
  onClick
}) {
  return <button type="button" title={(0, _Translate.translate)({
    id: 'theme.docs.sidebar.collapseButtonTitle',
    message: 'Collapse sidebar',
    description: 'The title attribute for collapse button of doc sidebar'
  })} aria-label={(0, _Translate.translate)({
    id: 'theme.docs.sidebar.collapseButtonAriaLabel',
    message: 'Collapse sidebar',
    description: 'The title attribute for collapse button of doc sidebar'
  })} className={(0, _clsx.default)('button button--secondary button--outline', _stylesModule.default.collapseSidebarButton)} onClick={onClick}>
      <_IconArrow.default className={_stylesModule.default.collapseSidebarButtonIcon} />
    </button>;
}

function DocSidebarDesktop({
  path,
  sidebar,
  onCollapse,
  isHidden
}) {
  const showAnnouncementBar = useShowAnnouncementBar();
  const {
    navbar: {
      hideOnScroll
    },
    hideableSidebar
  } = (0, _themeCommon.useThemeConfig)();
  const {
    isClosed: isAnnouncementBarClosed
  } = (0, _themeCommon.useAnnouncementBar)();
  return <div className={(0, _clsx.default)(_stylesModule.default.sidebar, {
    [_stylesModule.default.sidebarWithHideableNavbar]: hideOnScroll,
    [_stylesModule.default.sidebarHidden]: isHidden
  })}>
      {hideOnScroll && <_Logo.default tabIndex={-1} className={_stylesModule.default.sidebarLogo} />}
      <nav className={(0, _clsx.default)('menu thin-scrollbar', _stylesModule.default.menu, {
      [_stylesModule.default.menuWithAnnouncementBar]: !isAnnouncementBarClosed && showAnnouncementBar
    })}>
        <ul className={(0, _clsx.default)(_themeCommon.ThemeClassNames.docs.docSidebarMenu, 'menu__list')}>
          <_DocSidebarItem.DocSidebarItems items={sidebar} activePath={path} />
        </ul>
      </nav>
      {hideableSidebar && <HideableSidebarButton onClick={onCollapse} />}
    </div>;
}

const DocSidebarMobileSecondaryMenu = ({
  toggleSidebar,
  sidebar,
  path
}) => {
  return <ul className={(0, _clsx.default)(_themeCommon.ThemeClassNames.docs.docSidebarMenu, 'menu__list')}>
      <_DocSidebarItem.DocSidebarItems items={sidebar} activePath={path} onItemClick={() => toggleSidebar()} />
    </ul>;
};

function DocSidebarMobile(props) {
  return <_themeCommon.MobileSecondaryMenuFiller component={DocSidebarMobileSecondaryMenu} props={props} />;
}

const DocSidebarDesktopMemo = _react.default.memo(DocSidebarDesktop);

const DocSidebarMobileMemo = _react.default.memo(DocSidebarMobile);

function DocSidebar(props) {
  const windowSize = (0, _useWindowSize.default)(); // Desktop sidebar visible on hydration: need SSR rendering

  const shouldRenderSidebarDesktop = windowSize === 'desktop' || windowSize === 'ssr'; // Mobile sidebar not visible on hydration: can avoid SSR rendering

  const shouldRenderSidebarMobile = windowSize === 'mobile';
  return <>
      {shouldRenderSidebarDesktop && <DocSidebarDesktopMemo {...props} />}
      {shouldRenderSidebarMobile && <DocSidebarMobileMemo {...props} />}
    </>;
}
