// @docusaurus/theme-classic/lib/theme/LastUpdated/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = LastUpdated;

var _react = _interopRequireDefault(require("react"));

var _Translate = _interopRequireDefault(require("@docusaurus/Translate"));

var _themeCommon = require("@docusaurus/theme-common");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function LastUpdatedAtDate({
  lastUpdatedAt,
  formattedLastUpdatedAt
}) {
  return <_Translate.default id="theme.lastUpdated.atDate" description="The words used to describe on which date a page has been last updated" values={{
    date: <b>
            <time dateTime={new Date(lastUpdatedAt * 1000).toISOString()}>
              {formattedLastUpdatedAt}
            </time>
          </b>
  }}>
      {' on {date}'}
    </_Translate.default>;
}

function LastUpdatedByUser({
  lastUpdatedBy
}) {
  return <_Translate.default id="theme.lastUpdated.byUser" description="The words used to describe by who the page has been last updated" values={{
    user: <b>{lastUpdatedBy}</b>
  }}>
      {' by {user}'}
    </_Translate.default>;
}

function LastUpdated({
  lastUpdatedAt,
  formattedLastUpdatedAt,
  lastUpdatedBy
}) {
  return <span className={_themeCommon.ThemeClassNames.common.lastUpdated}>
      <_Translate.default id="theme.lastUpdated.lastUpdatedAtBy" description="The sentence used to display when a page has been last updated, and by who" values={{
      atDate: lastUpdatedAt && formattedLastUpdatedAt ? <LastUpdatedAtDate lastUpdatedAt={lastUpdatedAt} formattedLastUpdatedAt={formattedLastUpdatedAt} /> : '',
      byUser: lastUpdatedBy ? <LastUpdatedByUser lastUpdatedBy={lastUpdatedBy} /> : ''
    }}>
        {'Last updated{atDate}{byUser}'}
      </_Translate.default>
      {process.env.NODE_ENV === 'development' && <div>
          <small> (Simulated during dev for better perf)</small>
        </div>}
    </span>;
}
